riscv-elf/src/lib.rs

```diff
@@ -0,0 +1,379 @@
+        let header: &'elf ElfHeader64 = unsafe { &*(hbytes.as_ptr() as *const ElfHeader64) };
+        // Check magic number
+        if header.ei_magic != EI_MAGIC {
+            return Err(Error::InvalidMagicNumber);
+        }
+        // Check is 64bit ELF.
+        if header.ei_class != EI_CLASS_64 {
+            return Err(Error::InvalidClass);
+        }
+        // Check is Little-Endian
+        if header.ei_data != EI_DATA_LE {
+            return Err(Error::InvalidEndianness);
+        }
+        // Check ELF versions.
+        if header.ei_version != EI_VERSION_1 || header.e_version != E_VERSION_1 {
+            return Err(Error::BadElfVersion);
+        }
+        // Check is RISC-V.
+        if header.e_machine != E_MACHINE_RISCV {
+            return Err(Error::NotRiscV);
+        }
+
+        // Check Program Header Table
+        let phnum = header.e_phnum as usize;
+        let phentsize = header.e_phentsize as usize;
+        // Check that e_phentsize is >= of size of ElfProgramHeader64
+        if core::mem::size_of::<ElfProgramHeader64>() > phentsize {
+            return Err(Error::BadEntrySize);
+        }
+        // Check that we can read the program header table.
+        let program_headers =
+            slice_get_range(bytes, header.e_phoff, phnum * phentsize).ok_or(Error::BadOffset)?;
+
+        // Load segments
+        let mut segments = ArrayVec::<ElfSegment, ELF_SEGMENTS_MAX>::new();
+        let num_segs = core::cmp::min(phnum, ELF_SEGMENTS_MAX);
+        for i in 0..num_segs {
+            // Find the i-th ELF Program Header.
+            let phbytes = slice_get_range(program_headers, (i * phentsize).into(), phentsize)
+                .ok_or(Error::BadOffset)?;
+            // Safe because we are sure that the size of the slice is at least as big as ElfProgramHeader64
+            let ph: &'elf ElfProgramHeader64 =
+                unsafe { &*(phbytes.as_ptr() as *const ElfProgramHeader64) };
+
+            // Ignore if not a load segment.
+            if ph.p_type != PT_LOAD {
+                continue;
+            }
+
+            // Create a segment from the PH.
+            let datasz: usize = ph
+                .p_filesz
+                .try_into()
+                .map_err(|_| Error::ProgramHeaderMalformed)?;
+            let data = slice_get_range(bytes, ph.p_offset, datasz).ok_or(Error::BadOffset)?;
+            let vaddr = ph.p_vaddr;
+            let size: usize = ph
+                .p_memsz
+                .try_into()
+                .map_err(|_| Error::ProgramHeaderMalformed)?;
+            let flags = ph.p_flags;
+            let segment = ElfSegment::new(data, vaddr, size, flags)?;
+            segments.push(segment);
+        }
+
+        Ok(Self { segments })
+    }
+
+    /// Return an iterator containings loadable segments of this ELF file.
+    pub fn segments(&'elf self) -> impl Iterator<Item = &'elf ElfSegment> {
+        self.segments.iter()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I actually have a fuzzing infrastructure for this (super simple). But yeah, I can just take a random thing just to show some basic things in case someone changes these in the future.

Reply:
Wrote a few tests to check some (but definitely not all) of the cases of corrupted ELFs.